### DISCUSSION TITLE
Joining sessions to external transactions

### DISCUSSION THREAD

**Discussion by agronholm** [sqlalchemy/sqlalchemy#9013] [Usage Questions]
I'm trying to reproduce the test setup I used with SQLAlchemy 1.3, on 1.4. So far I've been unsuccessful and trying to figure out why. What I've found is that sessions behave very differently in regards to externally started transactions in v1.4. I reproduced the behavior in the script below. The problem I see is that each session releases the save point even though it didn't create it! This seems very, very wrong to me. A session should never release a save point unless the session itself created it. SQLAlchemy 1.3, from my POV, behaves correctly but 1.4 does not.

My test setup uses transactions on 3 different levels: session, module and test. Test level data is erased via rollback after each test, and module level test is seen by all tests within the same module. Test level data is seen by all tests/fixtures.

The [documentation](https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites) says that I should have a listener that recreates the save point when it's released or rolled back, but that totally destroys my use case. I have a cache class that, on cache miss, uses a throwaway session to fetch the missing data. This throwaway session is used because it would otherwise pollute the "main" session's identity map. Now think what happens when I close this throwaway session. It will either make all the changes made in the "main" session permanent on the module level (commit), or make it all vanish altogether (rollback), neither of which is tolerable.

<details>
<summary>Output with SQLAlchemy 1.3 (correct)</summary>

```
2022-12-20 13:35:10,380 INFO sqlalchemy.engine.base.Engine {}
Adding session data
2022-12-20 13:35:10,388 INFO sqlalchemy.engine.base.Engine INSERT INTO foo (name) VALUES (%(name)s)
2022-12-20 13:35:10,388 INFO sqlalchemy.engine.base.Engine {'name': 'session data'}
Adding module data
2022-12-20 13:35:10,389 INFO sqlalchemy.engine.base.Engine SAVEPOINT sa_savepoint_1
2022-12-20 13:35:10,389 INFO sqlalchemy.engine.base.Engine {}
2022-12-20 13:35:10,389 INFO sqlalchemy.engine.base.Engine INSERT INTO foo (name) VALUES (%(name)s)
2022-12-20 13:35:10,389 INFO sqlalchemy.engine.base.Engine {'name': 'module data'}
Adding test data
2022-12-20 13:35:10,390 INFO sqlalchemy.engine.base.Engine SAVEPOINT sa_savepoint_2
2022-12-20 13:35:10,390 INFO sqlalchemy.engine.base.Engine {}
2022-12-20 13:35:10,390 INFO sqlalchemy.engine.base.Engine INSERT INTO foo (name) VALUES (%(name)s)
2022-12-20 13:35:10,390 INFO sqlalchemy.engine.base.Engine {'name': 'test data'}
2022-12-20 13:35:10,391 INFO sqlalchemy.engine.base.Engine SELECT foo.name AS foo_name 
FROM foo
2022-12-20 13:35:10,391 INFO sqlalchemy.engine.base.Engine {}
3
2022-12-20 13:35:10,391 INFO sqlalchemy.engine.base.Engine ROLLBACK TO SAVEPOINT sa_savepoint_2
2022-12-20 13:35:10,391 INFO sqlalchemy.engine.base.Engine {}
2022-12-20 13:35:10,392 INFO sqlalchemy.engine.base.Engine ROLLBACK TO SAVEPOINT sa_savepoint_1
2022-12-20 13:35:10,392 INFO sqlalchemy.engine.base.Engine {}
2022-12-20 13:35:10,392 INFO sqlalchemy.engine.base.Engine ROLLBACK
```

</details>
<details>
<summary>Output with SQLAlchemy 1.4 (incorrect)</summary>

```
2022-12-20 13:35:48,171 INFO sqlalchemy.engine.Engine [no key 0.00004s] {}
Adding session data
2022-12-20 13:35:48,183 INFO sqlalchemy.engine.Engine INSERT INTO foo (name) VALUES (%(name)s)
2022-12-20 13:35:48,183 INFO sqlalchemy.engine.Engine [generated in 0.00010s] {'name': 'session data'}
Adding module data
2022-12-20 13:35:48,184 INFO sqlalchemy.engine.Engine SAVEPOINT sa_savepoint_1
2022-12-20 13:35:48,184 INFO sqlalchemy.engine.Engine [no key 0.00007s] {}
2022-12-20 13:35:48,184 INFO sqlalchemy.engine.Engine INSERT INTO foo (name) VALUES (%(name)s)
2022-12-20 13:35:48,184 INFO sqlalchemy.engine.Engine [cached since 0.001766s ago] {'name': 'module data'}
2022-12-20 13:35:48,185 INFO sqlalchemy.engine.Engine RELEASE SAVEPOINT sa_savepoint_1
2022-12-20 13:35:48,185 INFO sqlalchemy.engine.Engine [no key 0.00008s] {}
Adding test data
2022-12-20 13:35:48,185 INFO sqlalchemy.engine.Engine SAVEPOINT sa_savepoint_2
2022-12-20 13:35:48,185 INFO sqlalchemy.engine.Engine [no key 0.00006s] {}
2022-12-20 13:35:48,185 INFO sqlalchemy.engine.Engine INSERT INTO foo (name) VALUES (%(name)s)
2022-12-20 13:35:48,185 INFO sqlalchemy.engine.Engine [cached since 0.002937s ago] {'name': 'test data'}
2022-12-20 13:35:48,186 INFO sqlalchemy.engine.Engine RELEASE SAVEPOINT sa_savepoint_2
2022-12-20 13:35:48,186 INFO sqlalchemy.engine.Engine [no key 0.00006s] {}
2022-12-20 13:35:48,187 INFO sqlalchemy.engine.Engine SELECT foo.name AS foo_name 
FROM foo
2022-12-20 13:35:48,187 INFO sqlalchemy.engine.Engine [generated in 0.00006s] {}
3
/tmp/venv/test.py:54: SAWarning: nested transaction already deassociated from connection
  test_tx.rollback()
/tmp/venv/test.py:55: SAWarning: nested transaction already deassociated from connection
  module_tx.rollback()
2022-12-20 13:35:48,187 INFO sqlalchemy.engine.Engine ROLLBACK
```

</details>

```python
from sqlalchemy import Column, Unicode, create_engine
from sqlalchemy.orm import Session

try:
    from sqlalchemy.orm import declarative_base
except ImportError:
    from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class Foo(Base):
    __tablename__ = "foo"

    name = Column(Unicode, primary_key=True)

engine = create_engine("postgresql+psycopg2://postgres:secret@localhost:5432/eb_test", echo=True)
conn = engine.connect()
session_tx = conn.begin()

# Create tables within the session leveltransaction
Base.metadata.create_all(conn)

# Add test session level base data
print("Adding session data")
session = Session(conn)
session.add(Foo(name="session data"))
session.commit()
session.close()

# Create module level savepoint and add module level base data
print("Adding module data")
module_tx = conn.begin_nested()
session = Session(conn)
session.add(Foo(name="module data"))
session.commit()
session.close()

# Create test level savepoint and add base data
print("Adding test data")
test_tx = conn.begin_nested()
session = Session(conn)
session.add(Foo(name="test data"))
session.commit()
session.close()

# This should see all 3 Foo instances
session = Session(conn)
objects = session.query(Foo).all()
session.close()

print(len(objects))
assert len(objects) == 3

test_tx.rollback()
module_tx.rollback()
session_tx.rollback()
conn.close()
engine.dispose()
```

**Comment by zzzeek** [comment]
I haven't run the examples yet, but 1.3's version of this pattern requires a listener for the savepoints as well:

https://docs.sqlalchemy.org/en/13/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites

~~I'm not quite seeing~~ how an event listener "destroys the use case", you can add a listener to any session at any time, the listener here is used to re-start a new savepoint.   it's not new behavior at all that the Session will end the savepoint, at least for a rollback.   edit: OK it's nothing to do with the "listener"

**Reply by agronholm** [reply]
> it's not new behavior at all that the Session will end the savepoint, at least for a rollback.

But when simply closed, a session will release the savepoint on 1.4, but not on 1.3, as demonstrated in the output above. Rollback behavior is the same for both.

I'm trying to make this work on v1.4 by having those throwaway sessions start nested transactions, but it seems like a waste to do that when they get their own dedicated connections.

**Reply by zzzeek** [reply]
So you are using a classical SQLAlchemy pattern that is deprecated in SQLAlchemy 1.4 and is removed entirely in SQLAlchemy 2.0, which is the [Nesting of transaction blocks](https://docs.sqlalchemy.org/en/14/core/connections.html#nesting-of-transaction-blocks) feature.     The Session in 1.3 just calls begin() on the connection, which the connection then transparently handles as a savepoint, since it's already in "savepoint" mode.  it nests arbitrarily for each begin() call, and the nesting is only cancelled if rollback() is called, that's what the event listener is for.  your example only uses commit(), so the nesting in 1.3 is maintained.

Since you're a very experienced SQLAlchemy user, you actually used this pattern correctly, as I would also; when I implemented this feature, the "nesting" behavior felt intuitive and I wrote a bunch of docs about it, with the impression that this would be something widely used.

However, that's not how things turned out.  In practice, almost nobody was ever aware of this pattern, your example above might be the first I've ever seen of someone using it effectively.

The docs went unread by most, and to the degree that I saw code that used these features (in Openstack), it was not used effectively and led to just more confusion about how the connection / transaction worked.  In Openstack, more specifically they were stuck on the "nesting" pattern as applied to the Session and they had session.begin(subtransactions=True) hardcoded everywhere, the state of transactions in their apps was completely chaotic, and literally nobody understood what it meant or did.  I observed that most people working on Openstack (all professional developers who were paid to write code) hardly understood database transactions at all, ORM transactions were a total mystery to them, and then once you had "subtransactions" terminology, the vast majority of devs on Openstack had no idea what anything did so much that I was able to get hired to clean it all up.    I spent a few years getting them away from that pattern and wrote new libraries that adapted their code to use sessions / connections without any implicit "nesting".   

The failure of "arbitrary , implicit nesting" was clear to me, in that it was difficult to reason about, in an area of programming that is already deeply mysterious to most programmers, so I decided it had to go: https://gist.github.com/zzzeek/6af1e602fd8eb991564eeae7c9a39cb4#file-gistfile1-txt-L179    Which went great as in 2.0 the internal architecture of Connection is simpler, easier to understand, and much more robust.

the change in behavior you see is because `Session` is updated to work with a "future" mode connection that has no nesting behavior.    Whereas it previously called `begin()`, it now looks and sees that a nested transaction is present and just uses that as the current transaction.   what it does is coded directly to the pattern at https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites , and the case you present above was unknown (as it's a classical "nesting" pattern, which I've never seen anyone actually use.   SQLAlchemy 1.4 has been out for two years and no other issues have been raised).

the patch that changes the behavior in 1.4 to work just like 1.3:


```diff
diff --git a/lib/sqlalchemy/orm/session.py b/lib/sqlalchemy/orm/session.py
index 7d32362b50..6ed159475d 100644
--- a/lib/sqlalchemy/orm/session.py
+++ b/lib/sqlalchemy/orm/session.py
@@ -756,13 +756,16 @@ class SessionTransaction(TransactionalContext):
             elif self.nested:
                 transaction = conn.begin_nested()
             elif conn.in_transaction():
+                # use classical nesting
+                transaction = conn.begin()
+
                 # if given a future connection already in a transaction, don't
                 # commit that transaction unless it is a savepoint
-                if conn.in_nested_transaction():
-                    transaction = conn.get_nested_transaction()
-                else:
-                    transaction = conn.get_transaction()
-                    should_commit = False
+                # if conn.in_nested_transaction():
+                #     transaction = conn.get_nested_transaction()
+                # else:
+                #     transaction = conn.get_transaction()
+                #     should_commit = False
             else:
                 transaction = conn.begin()
         except:

```

this will fail if future=True on the engine as there's no nesting:

```
sqlalchemy.exc.InvalidRequestError: This connection has already initialized a SQLAlchemy Transaction() object via begin() or autobegin; can't call begin() here unless rollback() or commit() is called first.
```

another approach that modifies how this works, is that Session sees there's a nested transaction, OK, let's make yet another nested transaction; this is logically similar to how it worked in 1.3, except it's making real nested transactions, not "sub" transactions:

```diff
diff --git a/lib/sqlalchemy/orm/session.py b/lib/sqlalchemy/orm/session.py
index 7d32362b50..5ccff88c37 100644
--- a/lib/sqlalchemy/orm/session.py
+++ b/lib/sqlalchemy/orm/session.py
@@ -759,7 +759,8 @@ class SessionTransaction(TransactionalContext):
                 # if given a future connection already in a transaction, don't
                 # commit that transaction unless it is a savepoint
                 if conn.in_nested_transaction():
-                    transaction = conn.get_nested_transaction()
+                    transaction = conn.begin_nested()
+                    #transaction = conn.get_nested_transaction()
                 else:
                     transaction = conn.get_transaction()
                     should_commit = False
```

with that one, you will see lots of savepoints since it's making one on each go.  This would be the closest we could get to what you refer to as "correct", since we don't have a "subtransaction" anymore (technically we do, in the case of the "flush" process, however, otherwise reliance on "subtransactions" is something I want to stay away from, if possible).    however, right now Session's behavior, and we can document this more, is that joining the Session onto a savepoint means that Session will take over that savepoint.    I dont know that this is a time that this default behavior could be changed as we are on the cusp of 2.0 and generally I dont think we should have to since there's other ways to do this.     I can see that the above behavior might be something that should be possible, that is, to inject directly inside of there how the transaction is started.    that would require though that we really make a much more formal statement about what's happening in there and lock it down basically forever.

for the moment, we can show how to achieve the goal without making that change -  you've made some savepoints on the outside and you don't want them touched.  We can do that by approximating what the above patch does, which is to just make a third level of savepoint.  this allows that, as well as supporting commit/rollback:

```py
def test_session(conn):
    session = Session(conn)

    nested = conn.begin_nested()

    from sqlalchemy import event

    @event.listens_for(session, "after_transaction_end")
    def end_savepoint(session, transaction):
        nonlocal nested
        if not nested.is_active:
            nested = conn.begin_nested()

    return session
```

there's only one problem which is that we aren't closing out that "nested" when you're done using a Session, and this leads to warnings when closing out `test_tx` and `module_tx`.   So some mechanism to close out `nested` would be needed, such as :

```py
import contextlib
from sqlalchemy import event

@contextlib.contextmanager
def test_session(conn):
    nested = conn.begin_nested()

    with Session(conn) as session:
        @event.listens_for(session, "after_transaction_end")
        def end_savepoint(session, transaction):
            nonlocal nested
            if not nested.is_active:
                nested = conn.begin_nested()

        yield session

    nested.rollback()
```

then you have code like:

```py
test_tx = conn.begin_nested()
with test_session(conn) as session:
    session.add(Foo(name="test data"))
    session.commit()

```

That's the best I can do for the moment.    If we devise some way to plug that extra "begin_nested()" into the Session itself, so that it manages itself more cleanly, then you wouldn't have to manage that "nested" element on the outside.  however that's how it's designed to work at the moment.

**Reply by zzzeek** [reply]
OK it is very possible I can make this change for 2.0, all current tests seem to pass, so the additional SAVEPOINT does not seem to get in the way of anything:

```py
diff --git a/lib/sqlalchemy/orm/session.py b/lib/sqlalchemy/orm/session.py
index 7d32362b50..6d8d5a88a5 100644
--- a/lib/sqlalchemy/orm/session.py
+++ b/lib/sqlalchemy/orm/session.py
@@ -757,9 +757,11 @@ class SessionTransaction(TransactionalContext):
                 transaction = conn.begin_nested()
             elif conn.in_transaction():
                 # if given a future connection already in a transaction, don't
-                # commit that transaction unless it is a savepoint
+                # commit that transaction unless it is a savepoint, in
+                # which case we make our own savepoint so that we don't
+                # commit the one given.
                 if conn.in_nested_transaction():
-                    transaction = conn.get_nested_transaction()
+                    transaction = conn.begin_nested()
                 else:
                     transaction = conn.get_transaction()
                     should_commit = False

```

test support could be added to ensure the outer transaction is not committed.

for 1.4, we're locked down and can't make big changes like that, so would need to be a 2.0 only thing

**Reply by agronholm** [reply]
Thank you for taking the time to explain this in depth. I didn't know I was using a deprecated feature. I blogged about this style of testing back in 2013 ([here](https://alextechrants.blogspot.com/2013/08/unit-testing-sqlalchemy-apps.html)), albeit for just one level of nested transactions.

I will read through this properly later and resume my experiments then.

**Reply by zzzeek** [reply]
#9013

**Reply by lelit** [reply]
The issue is actually #9015

**Reply by zzzeek** [reply]
ouch my synesthesia got in the way on that , thanks

**Comment by zzzeek** [comment]
this improvement might be actually a game changer.  with that one line, it looks like we can do away with the need for event handler **in all cases**.  Seems a little hard to believe.   Here's a test case that works:

```py
from sqlalchemy import Column, Unicode, create_engine
from sqlalchemy.orm import Session
from sqlalchemy import event

engine = create_engine("postgresql+psycopg2://scott:tiger@localhost/test", echo=True)

conn = engine.connect()
trans = conn.begin()
nested = nn = conn.begin_nested()

session = Session(conn)
session.connection()
session.commit()

session.connection()
session.rollback()

session.begin_nested()
session.connection()
session.rollback()
session.rollback()

session.begin()
session.begin_nested()
session.commit()
session.rollback()


print("-------------")
assert nested is nn
assert nested.is_active
nested.rollback()
trans.rollback()

```

the above is not possible in 1.3 or 1.4 at all; 1.3 raises errors about subtransactions too.   they can "work" by using event handlers that reopen the transaction, but that still affects the outer "nested" variable.

only with this proposed change does "nested" end up being completely untouched.   will need to write some tests to confirm this is what I'm seeing.

great change!
